### PR TITLE
Heartbeat API, test and logging improvements

### PR DESCRIPTION
**Pull request by cangencer**
- Rename usages of heartBeat/HeartBeat to heartbeat/Heartbeat
- Change heartBeatStopped event to heartbeatResumed for better semantics
- Add logging for last heartbeat received/requested for better diagnostics
- Improve heartbeat tests to cover missing cases.

### CODE REVIEW

**Comment by mustafaiman** [comment]
👍

File: hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java [hunk 1]
```diff
@@ -269,21 +271,32 @@ public String getCloseReason() {
...
-    void heartBeatingSucceed() {
-        heartBeating = true;
-        lastHeartbeatMillis = System.currentTimeMillis();
+    void heartbeatResumed() {
+        isHeartBeating = true;
     }
 
-    public long getLastHeartbeatMillis() {
-        return lastHeartbeatMillis;
+    void heartbeatReceived() {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java]: this name doesn't explain if it is an 'getter' or 'do something'. Probably using 'onHeartBeatFailed' would be more descriptive.

**Comment by pveentjer** [comment]
Looks good. But this should be moved to 3.8 I guess since we are at the point of closing the 3.7 release. @jerrinot wdyt

**Comment by jerrinot** [comment]
given the nature of changes I am fine with this even in 3.7. 
:+1:

File: hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java (hunk 1 above)
Comment (cangencer) [inline hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java]: I've renamed them to "onHeartbeat.." to indicate they're event handlers.

**Comment by pveentjer** [comment]
👍 

merge once green

**Comment by Donnerbart** [comment]
:+1:

**Pull request merged**